include/zephyr/drivers/clock_control/clock_control_mcux_ccm_rev3.h

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2023 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_CLOCK_CONTROL_MCUX_CCM_H_
+#define ZEPHYR_INCLUDE_DRIVERS_CLOCK_CONTROL_MCUX_CCM_H_
+
+#include <stdint.h>
+#include <zephyr/device.h>
+
+#define IMX_CCM_MAX_SOURCES 4
+
+#define IMX_CCM_FIXED_FREQ(src) ((src).source.fixed.freq)
+#define IMX_CCM_PLL_MAX_FREQ(src) ((src).source.pll.max_freq)
+#define IMX_CCM_RATE_LIMIT(src) \
+	((src).type == IMX_CCM_TYPE_FIXED ? \
+	 IMX_CCM_FIXED_FREQ(src) : IMX_CCM_PLL_MAX_FREQ(src))
+
+enum imx_ccm_clock_state {
+	IMX_CCM_CLOCK_STATE_INIT = 0, /* initial state - clock may be gated or not */
+	IMX_CCM_CLOCK_STATE_GATED, /* clock is gated */
+	IMX_CCM_CLOCK_STATE_UNGATED, /* clock is not gated */
+};
+
+enum imx_ccm_type {
+	IMX_CCM_TYPE_FIXED = 1,
+	IMX_CCM_TYPE_PLL,
+	IMX_CCM_TYPE_MAX,
+};
+
+struct imx_ccm_fixed {
+	char *name;
+	uint32_t id;
+	uint32_t freq;
+};
+
+struct imx_ccm_pll {
+	char *name;
+	uint32_t id;
+	uint32_t max_freq; /* maximum allowed frequency */
```

Comment:
Hm, configuring the PLLs is somewhat delicate IMO as if it's done during runtime we need to take into account the fact that other subsystems may be affected. I'd say it would be easier to configure them during driver initialization (possibly with the help of some devicetree properties). Either way, the way they are configured should be entirely up to the SoC developer. There are no restrictions ATM regarding this.